src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.DocumentHighlighting;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.NavigateTo;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Protocol.LanguageServices.Extensions;
+using Microsoft.CodeAnalysis.QuickInfo;
+using Microsoft.CodeAnalysis.Structure;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using VSLocation = Microsoft.VisualStudio.LanguageServer.Protocol.Location;
+
+namespace Microsoft.CodeAnalysis.Protocol.LanguageServices
+{
+    /// <summary>
+    /// This is the roslyn language service.
+    ///
+    /// This glues the LSP requests and solution to 
+    /// Roslyn (and services) and outputs LSP types.
+    /// </summary>
+    public sealed class RoslynLanguageService
+    {
+        // for now, it is null but later we might use this to provide info on types on dlls
+        private readonly IMetadataAsSourceFileService _metadataAsSourceService = null;
+
+        // TODO - Move hierarchicalDocumentSymbolSupport to client capabilities.
+        // https://github.com/dotnet/roslyn/projects/45#card-20033973
+        private readonly bool _hierarchicalDocumentSymbolSupport;
+        private readonly ClientCapabilities _clientCapabilities;
+
+        public RoslynLanguageService(ClientCapabilities clientCapabilities, bool hierarchicalDocumentSymbolSupport)
+        {
+            _hierarchicalDocumentSymbolSupport = hierarchicalDocumentSymbolSupport;
+            _clientCapabilities = clientCapabilities;
+        }
+
+        public ServerCapabilities GetServerCapabilities()
+        {
+            return new ServerCapabilities
+            {
+                DefinitionProvider = true,
+                ReferencesProvider = true,
+                DocumentHighlightProvider = true,
+                DocumentSymbolProvider = true,
+                WorkspaceSymbolProvider = true,
+                HoverProvider = true,
+            };
+        }
+
+        /// <summary>
+        /// Answers a document symbols request by returning a list of symbols in the document.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol
+        /// </summary>
+        /// <param name="solution">the solution containing the document.</param>
+        /// <param name="request">the document to get symbols from.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of symbols in the document.</returns>
+        public async Task<object[]> GetDocumentSymbolsAsync(Solution solution, DocumentSymbolParams request, CancellationToken cancellationToken)
```

Comment:
fwiw, i would far prefer each of these different, complex, capabilities be broken out into separate files (and ideally separate classes).  just having this huge file for everything isn't desirable.